File: packages/service-job-js/src/core/utils/DataLoader/DataLoader.ts
import { NotFoundError } from '@lib/shared/core/errors/NotFoundError/NotFoundError';
import { Container } from '@lib/shared/core/utils/Container/Container';
import { type SourcedEntityResourceModel } from 'packages/service-job-js/src/core/core.models';
import {
  type DataLoaderModel,
  type DataLoaderParamsModel,
} from 'packages/service-job-js/src/core/utils/DataLoader/DataLoader.models';

export class DataLoader<TType extends SourcedEntityResourceModel>
  implements DataLoaderModel<TType>
{
  params!: DataLoaderParamsModel<TType>;

  constructor(params: DataLoaderParamsModel<TType>) {
    this.params = params;
  }

  async fetch(): Promise<Array<Partial<TType>>> {
    throw new Error('Method not implemented.');
  }

  async fetchPostProcess(): Promise<Array<Partial<TType>>> {
    return (await this.fetch()).map((v) => ({ ...v, source: v.source ?? this.params?.source }));
  }

  async upload(): Promise<Array<Partial<TType>>> {
    if (this.params.ResourceImplementation) {
      const data = await this.fetchPostProcess();
      await Container.get(this.params.ResourceImplementation)?.createMany?.({ form: data });
      return data;
    }
    throw new NotFoundError('resource implementaiton');
  }

  get source(): string | undefined {
    return this.params.source;
  }
}

File: packages/lib-shared-py/src/lib_shared/database/utils/data_loader/data_loader.py
from typing import Generic, TypeVar

from lib_model.core.utils.database_entity.database_entity import DatabaseEntity

from lib_shared.core.utils.base_model.base_model import BaseModel
from lib_shared.core.utils.field.field import Field
from lib_shared.core.utils.logger.logger import logger
from lib_shared.database.utils.database import Database

TType = TypeVar("TType", bound=DatabaseEntity)


class DataLoader(BaseModel, Generic[TType]):
    resource: DatabaseEntity
    source: str | None = None
    database: Database = Field(default_factory=lambda: db)

    async def load(
        self,
    ) -> list[TType]:
        raise NotImplementedError("load() method not implemented")

    async def upload(
        self,
    ) -> list[TType]:
        data = await self.load()
        if len(data):
            if self.database:
                if self._params.source:
                    for v in data:
                        v.source = v.source or self._params.source

                result = await self._params.database.create_many(
                    data=data,
                    resource=self._params.resource,
                )
                return result.result
            raise Exception("Database not initialized")
        else:
            logger.info("no data to upload")
            return []
